Better example (#39)
cose-wg/draft-ietf-cose-hash-envelope: draft-ietf-cose-hash-envelope.md

# Envelope EDN

> The following informative example demonstrates how to construct a hash envelop for a resource which is already commonly referenced by its hash.

Reviewer: ```suggestion
The following informative example demonstrates how to construct a hash envelope for a resource already commonly referenced by its hash.
```

~~~~ cbor-diag
18([ # cose-sign1
  <<{
    / signature algorithm / 1: -35, # ES384
    / key identifier      / 4: h'75726e3a...32636573',
    / cose sign1 type     / 16: "application/example+cose",
    / hash algorithm      / TBD_1: -16, # sha256
    / media type          / TBD_2: "application/spdx+json",
    / location            / TBD_3: "https://sbom.example/" + ... + "/manifest.spdx.json"
  }>>
  / unprotected / {},
  / payload     / h'935b5a91...e18a588a', # As seen in manifest.spdx.json.sha256

>   / signature   / h'15280897...93ef39e5'  # ECDSA Signature with SHA 384 and P-384

Reviewer: this reformatting doesn't fit in the 70 characters. 
I'll do a quick fixup...,

])
~~~~

> In this example, an spdx software bill of materials (sbom) in json format is already commonly identified with its sha256 hash function, for example many tools will generate a file called `manifest.spdx.json.sha256` which contains that sha256 hash of the `manifest.spdx.json`.

Reviewer 2: ```suggestion
In this example, an SPDX software bill of materials (SBOM) in JSON format is already commonly identified by a SHA256 hash function. For example, some tooling generates a file, such as `manifest.spdx.json.sha256`, which contains the SHA256 hash of the corresponding `manifest.spdx.json` file.
```

The content type for `manifest.spdx.json` is already well known as `application/spdx+json`, and is registered with IANA [here](https://www.iana.org/assignments/media-types/application/spdx+json).

The full JSON SBOM is available at a URL, such as `https://sbom.example/.../manifest.spdx.json`.

> The payload of this cose-sign1 is the SHA256 hash of the `manifest.spdx.json`, which is sometimes found in an adjacent file called `manifest.spdx.json.sha256`.

Reviewer 2: ```suggestion
The payload of this COSE_Sign1 is the SHA256 hash of the `manifest.spdx.json`, which is typically found in an adjacent file, e.g., `manifest.spdx.json.sha256`.
```

> The payload of this cose-sign1 is the sha256 hash of the `manifest.spdx.json`, which is sometimes found in an adjacent file called `manifest.spdx.json.sha256`.

Reviewer 2: ```suggestion
The payload of such COSE_Sign1 are the SHA256 hashes of the content data in files, such as  `manifest.spdx.json`. Such COSE_Sign1 documents are typically located adjacent corresponding SHA256 hash files, e.g., `manifest.spdx.json.sha256`.
```

Tried to fix, but what? the payload of the COSE envelope is somtimes found in the hash file? Maybe I am reading this wrong.

> The payload of this cose-sign1 is the sha256 hash of the `manifest.spdx.json`, which is sometimes found in an adjacent file called `manifest.spdx.json.sha256`.

Author: ```suggestion
The payload of this cose-sign1 is the SHA256 hash of the `manifest.spdx.json`, which is sometimes found in an adjacent file called `manifest.spdx.json.sha256`.
```

> The type of this cose-sign1 is `application/example+cose`, but other types may be used to establish more specific media types for signatures of hashes.

Reviewer 2: ```suggestion
The type of this SHA256 hash is `application/example+cose`, but other types may be used to establish more specific media types for signatures of hashes.
```

The signature is produced using ES384 which means using ECDSA with SHA384 hash function and P-384 elliptic curve.

This example is chosen to highlight that an existing system may use a hash algorithm such as sha256.
This hash becomes the payload of a cose-sign1.

> When signed with a signature algorithm that is paramaterized by hash function, such as ECDSA with SHA384, the to be signed structure as described in Section 4.4 of RFC9052.

Reviewer 2: ```suggestion
When signed with a signature algorithm that is parameterized via a hash function, such as ECDSA with SHA384, the to be signed structure as described in Section 4.4 of RFC9052.
```

This sentence no verb.

Reviewer: >This sentence no verb.
Says the german to the american....

> The resulting signature is over the protected header and payload, providing integrity and authenticity for the hash algorithm, content type and location of the associated resource, in this case a software bill of materials.

Reviewer 2: ```suggestion
The resulting signature is computed over the protected header and payload, providing integrity and authenticity for the hash algorithm, content type and location of the associated resource, in this case a software bill of materials.
```

Isn't location allowed to be in the unprotected header now?

Reviewer: > Isn't location allowed to be in the unprotected header now? 
👎 

However, the updated text doesn't reflect one way or another.

# Envelope EDN

The following informative example demonstrates how to construct a hash envelope for a resource already commonly referenced by its hash.

~~~~ cbor-diag
18([ # cose-sign1
  <<{
    / signature alg   / 1: -35, # ES384
    / key identifier  / 4: h'75726e3a...32636573',
    / cose sign1 type / 16: "application/example+cose",
    / hash algorithm  / TBD_1: -16, # sha256
    / media type      / TBD_2: "application/spdx+json",
    / location        /
         TBD_3: "https://sbom.example/.../manifest.spdx.json"
  }>>
  / unprotected / {},
  / payload     / h'935b5a91...e18a588a',
         # As seen in manifest.spdx.json.sha256
  / signature   / h'15280897...93ef39e5'
         # ECDSA Signature with SHA 384 and P-384
])
~~~~

In this example, an SPDX software bill of materials (SBOM) in JSON format is already commonly identified by its SHA256 hash.
For example, some tooling generates a file, such as `manifest.spdx.json.sha256`, which contains the SHA256 hash of the corresponding `manifest.spdx.json` file.

> The content type for `manifest.spdx.json` is already well known as `application/spdx+json`, and is registered with IANA [here](https://www.iana.org/assignments/media-types/application/spdx+json).

Reviewer: ```suggestion
The content type for `manifest.spdx.json` is already well known as `application/spdx+json`, and is [registered with IANA](https://www.iana.org/assignments/media-types/application/spdx+json).
```

The full JSON SBOM is available at a URL, such as `https://sbom.example/.../manifest.spdx.json`.

> The payload of this COSE_Sign1 is the SHA256 hash of the `manifest.spdx.json`, which is typically found in an adjacent file, e.g., `manifest.spdx.json.sha256`.

Reviewer: ```suggestion
The payload of this COSE_Sign1 is the SHA256 hash of the `manifest.spdx.json`, which is typically found in an adjacent file (`manifest.spdx.json.sha256`).
```

The type of this COSE_Sign1 is `application/example+cose`, but other types may be used to establish more specific media types for signatures of hashes.

The signature is produced using ES384 which means using ECDSA with SHA384 hash function and P-384 elliptic curve.

This example is chosen to highlight that an existing system may use a hash algorithm such as sha256.

> This hash becomes the payload of a cose-sign1.

Reviewer: ```suggestion
This hash becomes the payload of a COSE-Sign1.
```

- Label: TBD_1
- Value type: int
- Value registry: https://www.iana.org/assignments/cose/cose.xhtml#algorithms

> - Description: The hash algorithm used to produce the payload of a COSE_Sign1.

Author: ```suggestion
- Description: Hash algorithm used to produce the payload from pre-image content
```

- Name: payload_preimage_content_type
- Label: TBD_2
- Value type: uint / tstr

> - Value registry: [COAP Content-Formats](https://www.iana.org/assignments/core-parameters/core-parameters.xhtml#content-formats) or [Media Types](https://www.iana.org/assignments/media-types/media-types.xhtml) registry.
> - Description: The content format or media type of data that has been hashed to produce the payload of a COSE_Sign1.

Author: ```suggestion
Value registry when `uint` is used: https://www.iana.org/assignments/core-parameters/core-parameters.xhtml#content-formats
- Description: The content format associated with the bytes that were hashed to produce the payload.
  `uint` payload_preimage_content_types SHOULD be registered in the content-formats registry.
  `tstr` values MAY be used when registered values may not yet be registered.
```

- Name: payload_preimage_content_type
- Label: TBD_2
- Value type: uint / tstr

> Value registry when `uint` is used: https://www.iana.org/assignments/core-parameters/core-parameters.xhtml#content-formats

Author: ```suggestion
- Value registry when `uint` is used: https://www.iana.org/assignments/core-parameters/core-parameters.xhtml#content-formats
```

- Label: TBD_3
- Value type: tstr
- Value registry: none

> - Description: The string or URI hint for the location of the data hashed to produce the payload of a COSE_Sign1.

Author: ```suggestion
- Description: A string or URI as a hint for the location of the payload
```
